src/legacy/ui/public/kfetch/kfetch.ts

```diff
@@ -48,32 +47,36 @@ const interceptors: Interceptor[] = [];
 export const resetInterceptors = () => (interceptors.length = 0);
 export const addInterceptor = (interceptor: Interceptor) => interceptors.push(interceptor);
 
-export async function kfetch(
-  options: KFetchOptions,
-  { prependBasePath = true }: KFetchKibanaOptions = {}
-) {
-  const combinedOptions = withDefaultOptions(options);
-  const promise = requestInterceptors(combinedOptions).then(
-    ({ pathname, query, ...restOptions }) => {
-      const fullUrl = url.format({
-        pathname: prependBasePath ? chrome.addBasePath(pathname) : pathname,
-        query,
-      });
-
-      return window.fetch(fullUrl, restOptions).then(async res => {
-        if (!res.ok) {
-          throw new KFetchError(res, await getBodyAsJson(res));
-        }
-        const contentType = res.headers.get('content-type');
-        if (contentType && contentType.split(';')[0] === 'application/ndjson') {
-          return await getBodyAsBlob(res);
-        }
-        return await getBodyAsJson(res);
-      });
-    }
-  );
+export function createKfetch(http: HttpSetup) {
+  return function kfetch(
+    options: KFetchOptions,
+    { prependBasePath = true }: KFetchKibanaOptions = {}
+  ) {
+    const controller = new AbortController();
+    const promise = new Promise((resolve, reject) => {
+      responseInterceptors(
+        requestInterceptors(withDefaultOptions(options))
+          .then(({ pathname, query, ...restOptions }) =>
+            http.fetch(pathname, {
+              ...restOptions,
+              signal: controller.signal,
+              query,
+              prependBasePath,
+            })
+          )
+          .catch(err => {
+            throw new KFetchError(err.response || { statusText: err.message }, err.body);
+          })
+      ).then(resolve, reject);
```

Comment:
just curious -why we need to wrap in another promise?

Reply:
Let's say you return a promise which captures the abortable promise along the way:

```js
const promise = Promise
  .reject(example)
  .then(() => http.fetch(path, { signal });
```

`promise.abort()` does not exist on the promise because it is only on the promise returned by `http.fetch()`. If an error occurs along the way and that promise isn't returned, then this would throw.

All that is to say, we need an abortable promise to exist from the very beginning to avoid an exception from this missing method.